shotover-proxy/src/codec/cassandra.rs

```diff
@@ -31,20 +33,58 @@ impl Default for CassandraCodec {
 impl CassandraCodec {
     pub fn new() -> CassandraCodec {
         CassandraCodec {
-            compression: Compression::None,
+            compression: Arc::new(RwLock::new(Compression::None)),
             messages: vec![],
             current_use_keyspace: None,
         }
     }
 }
 
+impl Codec for CassandraCodec {
+    fn clone_without_state(&self) -> Self {
+        Self {
+            compression: Arc::new(RwLock::new(Compression::None)),
+            messages: self.messages.clone(),
+            current_use_keyspace: self.current_use_keyspace.clone(),
+        }
+    }
+}
+
 impl CassandraCodec {
-    fn encode_raw(&mut self, item: CassandraFrame, dst: &mut BytesMut) {
-        let buffer = item.encode(self.compression);
-        if buffer.is_empty() {
-            info!("trying to send 0 length frame");
+    fn check_compression(&mut self, bytes: &BytesMut) -> Result<bool> {
+        if bytes.len() < 9 {
+            return Err(anyhow!("Not enough bytes for cassandra frame"));
+        }
+        let opcode = Opcode::try_from(bytes[4])?;
+
+        let compressed = Flags::from_bits_truncate(bytes[1]).contains(Flags::COMPRESSION);
+
+        // check if startup message and set the codec's selected compression
+        if Opcode::Startup == opcode {
+            if let CassandraFrame {
+                operation: CassandraOperation::Startup(startup),
+                ..
+            } = CassandraFrame::from_bytes(bytes.clone().freeze(), Compression::None)?
+            {
+                self.set_compression(&startup);
+            };
+            return Ok(false);
```

Comment:
What is this early return for?
Cant we just rely on the compressed flag like we usually do?

Reply:
Startup should never be compressed.

Reply:
But even if it is compressed, pretending that it isn't won't save us.